File: mobile/sharedLogic/src/commonMain/kotlin/org/example/project/GreetingClient.kt
package org.example.project

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.client.request.parameter
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class GreetingResponse(val message: String)

class GreetingClient(
    private val baseUrl: String,
    private val httpClient: HttpClient = createHttpClient(),
) {
    suspend fun fetchGreeting(name: String): String {
        val response: GreetingResponse =
            httpClient
                .get("$baseUrl/api/greeting") {
                    parameter("name", name)
                }
                .body()
        return response.message
    }

    companion object {
        fun create(baseUrl: String = apiBaseUrl()): GreetingClient = GreetingClient(baseUrl)
    }
}

private fun createHttpClient(): HttpClient =
    HttpClient {
        install(ContentNegotiation) {
            json(
                Json {
                    ignoreUnknownKeys = true
                },
            )
        }
    }

File: web/src/api/greetingClient.ts
import { apiBaseUrl } from "@/config"
import type { GreetingResponse } from "@/api/types"

export class GreetingClient {
  private readonly baseUrl: string
  private readonly fetchFn: typeof fetch

  constructor(
    baseUrl: string = apiBaseUrl,
    fetchFn: typeof fetch = globalThis.fetch.bind(globalThis),
  ) {
    this.baseUrl = baseUrl
    this.fetchFn = fetchFn
  }

  async fetchGreeting(name: string): Promise<string> {
    const url = new URL("/api/greeting", ensureTrailingSlash(this.baseUrl))
    url.searchParams.set("name", name)

    const response = await this.fetchFn(url)
    if (!response.ok) {
      throw new Error(`Greeting request failed (${response.status})`)
    }

    const body = (await response.json()) as GreetingResponse
    if (typeof body.message !== "string") {
      throw new Error("Invalid greeting response")
    }
    return body.message
  }
}

function ensureTrailingSlash(baseUrl: string): string {
  return baseUrl.endsWith("/") ? baseUrl : `${baseUrl}/`
}
